cf-custom-resources/lib/backlog-per-task-calculator.js

```diff
@@ -0,0 +1,129 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+"use strict";
+
+const aws = require("aws-sdk");
+
+// AWS Clients that are overriden in tests.
+let ecs, sqs;
+
+/**
+ * This lambda function calculates the backlog of SQS messages per running ECS tasks,
+ * and writes the metric to CloudWatch.
+ */
+exports.handler = async (event, context) => {
+  setupClients();
+  try {
+    const runningCount = await getRunningTaskCount(process.env.CLUSTER_NAME, process.env.SERVICE_NAME);
+    const backlogs = await Promise.all(
+      convertQueueNames(process.env.QUEUE_NAMES).map(async (queueName) => {
+        const queueUrl = await getQueueURL(queueName);
+        return {
+          queueName: queueName,
+          backlogPerTask: await getBacklogPerTask(queueUrl, runningCount),
+        };
+      })
+    );
+    const timestamp = Date.now();
+    for (const {queueName, backlogPerTask} of backlogs) {
+      emitBacklogPerTaskMetric(process.env.NAMESPACE, timestamp, queueName, backlogPerTask);
+    }
+  } catch(err) {
+    // If there is any issue we won't log a metric.
+    // This is okay because autoscaling will maintain the current number of running tasks if a data point is missing.
+    // See https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-autoscaling-targettracking.html#targettracking-considerations
+    console.error(`Unexpected error ${err}`);
+  }
+};
+
+/**
+ * Returns the backlog per task. The backlog per task is calculated by dividing the number of messages in the queue with
+ * the number of running tasks.
+ * If there are no running task, we return the total number of messages in the queue so that we can start scaling up.
```

Comment:
> Could it be possible that the calculated acceptable backlog > total number of messages

Hmm I don't think this is possible (unless I am missing something 🤔 ) We can only end up in this scenario if `runningCount` from ECS returns a float :P between [0, 1). But I might be missing something.

> do we want to scale up from 0,

Yeah, that's the intention. If there are messages in the queue then I think we should scale up otherwise if empty remain at 0.